src/electron/windows/sign_windows_executable.mjs

```diff
@@ -0,0 +1,212 @@
+    jsignProc.on('error', reject);
+    jsignProc.on('exit', resolve);
+  });
+}
+
+/**
+ * Sign the target exeFile using a specific algorithm and options.
+ * @param {string} exeFile the full path of the exe file to be signed.
+ * @param {'sha1'|'sha256'} algorithm the algorithm used for signing.
+ * @param {object} options additional options (cli arguments) for signing.
+ *                         the options will also be read from environment
+ *                         variables.
+ */
+export async function signWindowsExecutable(exeFile, algorithm, options) {
+  const type = getOptionValue(options, 'certtype', 'WINDOWS_SIGNING_CERT_TYPE', false);
+  if (!type || type === 'none') {
+    console.info(`skip signing "${exeFile}"`);
+    return;
+  }
+
+  assert(!!exeFile, 'executable path is required');
+  assert(
+      algorithm === 'sha1' || algorithm === 'sha256',
+      'hashing algorithm must be either "sha1" or "sha256"');
+
+  exeFile = resolve(exeFile);
+  await assertFileExists(exeFile, 'executable file "%s" does not exist');
+
+  const password = getOptionValue(options, 'password', 'WINDOWS_SIGNING_CERT_PASSWORD', true);
+
+  var jsignArgs = [
+    '--alg', algorithm === 'sha256' ? 'SHA-256' : 'SHA-1', '--tsaurl',
+    'http://timestamp.digicert.com', '--storepass', password
+  ];
+
+  switch (type) {
+    case 'pfx':
+      appendPfxJsignArgs(jsignArgs, options);
+      break;
+    case 'digicert-usb':
+      appendDigicertUsbJsignArgs(jsignArgs, options);
+      break;
+    default:
+      throw new Error(`cert type ${type} is not supported`);
+  }
+
+  var exitCode;
+  try {
+    exitCode = await jsign(exeFile, jsignArgs);
+  } catch (err) {
+    console.error('failed to start java, please make sure you have installed java');
+    throw new Error(err);
+  }
+
+  if (exitCode === 0) {
+    console.info(`successfully signed "${exeFile}"`);
+  } else {
+    console.error(`jsign exited with error code ${exitCode}`);
+    throw new Error(`failed to sign "${exeFile}"`);
+  }
+}
+
+async function main() {
+  const {target, algorithm, ...options} = minimist(process.argv);
+  await signWindowsExecutable(target, algorithm, options);
+}
+
+// Call this script through CLI to sign a Windows executable:
+//   node sign_windows_executable.mjs
+//     --target <exe-path-to-sign>
+//     --algorithm <sha1|sha256>
+//     --certtype <none|pfx|digicert-usb>
+//     --password <cert-store-password>
+// The following options are for --certtype == pfx
+//     --pfx <pfx-cert-path>
+// The following options are for --certtype == digicert-usb
+//     [--subject <cert-subject-name>]
+//
+// You can also use environment variables to specify some arguments:
+//   WINDOWS_SIGNING_CERT_TYPE       <=> --certtype
+//   WINDOWS_SIGNING_CERT_PASSWORD   <=> --password
+//   WINDOWS_SIGNING_PFX_CERT        <=> --pfx
+//   WINDOWS_SIGNING_EV_CERT_SUBJECT <=> --subject
+if (import.meta.url === pathToFileURL(process.argv[1]).href) {
+  main();
```

Comment:
we should align on how we want to handle the top level execution at some point - https://stackoverflow.com/questions/46515764/how-can-i-use-async-await-at-the-top-level

Reply:
Sure, in the question you pasted, people wants to use `await` in the top level, for example:

```javascript
try {
  await main();
catch (err) {
  console.error(err);
}
```

This is not allowed and we have to wrap it in an anonymous `async` function:

```javascript
async function() {
  ... // same as above
}();
```

But in our case, we don't need to `await` anything here (node will do that if we are calling an `async` function, which is `main` in this case), therefore we can use `main()` directly, actually it is the same as `async function() {...}()` as well.

So my opinion would be never use `await` at the root level, and so we can simply call `main()` in the root level without any anonymous `async` function wrappers. But I'm open to other styles as well.

Reply:
To summarize, the issue here is largely about exposing and handling potential errors so they're not swallowed by the promise, and it might be good to do so at this root level rather than risking them be inconsistently handled otherwise.

I tried doing the top-level await before and it wasn't working, but maybe I had nvm accidentally set to older version of node. If you can confirm it works, `try { await main() } catch (err) { console.error(err) };` is great

Reply:
Yup, I tried `await main()` and errors are caught. Updated the code.